vertexy/src/private/constraints/ShortestPathConstraint.cpp

```diff
@@ -0,0 +1,596 @@
+			// source for the target)
+			lits.push_back(Literal(potentialSource, m_sourceMask));
+			visited[potentialSrcIndex] = true;
+			foundAny = true;
+		}
+	}
+	vxy_assert(foundAny);
+	if (isGreaterOp)
+	{
+		m_explanationMinGraph->fastForward();
+	}
+
+	if (hasUnprocessedSources)
+	{
+		// for all other sources, run the reachability explainer
+		auto otherLits = explainNoReachability(params, &visited);
+		// we can skip the first one since it's always the propagatedVariable
+		for (int i = 1; i < otherLits.size(); i++)
+		{
+			const auto& lit = otherLits[i];
+			if (edgeVarsRecorded.find(lit.variable) == edgeVarsRecorded.end())
+			{
+				// we don't need to add that edge to the edgeVarsRecorded because it was already deduped
+				lits.push_back(lit);
+			}
+		}
+	}
+
+	return lits;
+}
+
+void Vertexy::ShortestPathConstraint::createTempSourceData(ReachabilitySourceData& data, int vertexIndex) const
+{
+	ITopologySearchConstraint::createTempSourceData(data, vertexIndex);
+#if REACHABILITY_USE_RAMAL_REPS
+	data.minReachability = make_shared<RamalRepsType>(m_minGraph, false, false, false);
+#else
+	Data.minReachability = make_shared<ESTreeType>(MinGraph);
+#endif
+	data.minReachability->initialize(vertexIndex, &m_reachabilityEdgeLookup, m_totalNumEdges);
+}
+
+void ShortestPathConstraint::onVertexChanged(int vertexIndex, VarID sourceVar, bool inMinGraph)
+{
+	vxy_assert(!m_backtracking);
+	vxy_assert(!m_explainingSourceRequirement);
+
+	vxy_assert(m_edgeChangeDb != nullptr);
+	vxy_assert(m_inEdgeChange);
+
+	if (m_edgeChangeFailure)
+	{
+		// We already failed - avoid further failures that could confuse the conflict analyzer
+		return;
+	}
+
+	m_processingVertices = true;
+	m_queuedVertexChanges.insert(vertexIndex);
+	auto reachability = determineValidity(m_edgeChangeDb, vertexIndex);
+
+	// TODO understand why uncommenting both of these make everything run much, much slower
+	//if (reachability == EValidityDetermination::DefinitelyValid)
+	//{
+	//	// vertexIndex became unreachable in the max graph
+	//	VarID var = m_sourceGraphData->get(vertexIndex);
+
+	//	if (var.isValid() && m_edgeChangeDb->getPotentialValues(var).anyPossible(m_requireValidMask))
+	//	{
+	//		auto ret = m_edgeChangeDb->constrainToValues(var, m_requireValidMask, this);
+	//		vxy_assert(ret);
+	//	}
+	//}
+	//else if (reachability == EValidityDetermination::PossiblyValid)
+	//{
+	//	// vertexIndex became unreachable in the max graph
+	//	VarID var = m_sourceGraphData->get(vertexIndex);
+
+	//	m_edgeChangeDb->
+	//	if (var.isValid() && !m_edgeChangeDb->constrainToValues(var, m_invalidMask.inverted(), this))
+	//	{
+	//		m_edgeChangeFailure = true;
+	//	}
+	//}
+	if (reachability == EValidityDetermination::DefinitelyUnreachable)
+	{
+		// vertexIndex became unreachable in the max graph
+		VarID var = m_sourceGraphData->get(vertexIndex);
+		//sanityCheckUnreachable(db, vertexIndex);
```

Comment:
In this code in particular, adding a sanity check will break very frequently because we are still processing the RamalReps, so they aren't all updated at this point. I'll move that to the code we do after all of them are updated

Reply:
Oh man maybe that's why we are getting so many backtracks!?

Reply:
I'll make some tests